# Enforce async queries (`findBy*`, `findAllBy*`) to have proper `await` (await-async-query)

TODO: Please describe the origin of the rule here.

## Rule Details

This rule aims to...

Examples of **incorrect** code for this rule:

```js
// TODO: fill me in
```

Examples of **correct** code for this rule:

```js
// TODO: fill me in
```

### Options

TODO: If there are any options, describe them here. Otherwise, delete this section.

## When Not To Use It

TODO: Give a short description of when it would be appropriate to turn off this rule.

## Further Reading

- [Testing Library queries cheatsheet](https://testing-library.com/docs/dom-testing-library/cheatsheet#queries)
