# Disallow the use of `cleanup` (no-manual-cleanup)

TODO: SUMMARY

## Rule Details

TODO: DETAILS

Examples of **incorrect** code for this rule:

```js
// TODO: EXAMPLES
```

Examples of **correct** code for this rule:

```js
// TODO: EXAMPLES
```

## Further Reading

- [cleanup API in React Testing Library](https://testing-library.com/docs/react-testing-library/api#cleanup)
